internal/lint/src/main/java/com/firebaseui/lint/FirestoreRecyclerAdapterLifecycleDetector.kt

```diff
@@ -0,0 +1,176 @@
+package com.firebaseui.lint
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity.WARNING
+import org.jetbrains.uast.UCallExpression
+import org.jetbrains.uast.UClass
+import org.jetbrains.uast.UField
+import org.jetbrains.uast.visitor.AbstractUastVisitor
+import java.util.EnumSet
+
+class FirestoreRecyclerAdapterLifecycleDetector : Detector(), Detector.UastScanner {
+
+  override fun getApplicableUastTypes() = listOf(UClass::class.java)
+
+  override fun createUastHandler(context: JavaContext) = MissingLifecycleMethodsVisitor(context)
+
+  class MissingLifecycleMethodsVisitor(
+      private val context: JavaContext
+  ) : UElementHandler() {
+    private val FIRESTORE_RECYCLER_ADAPTER_TYPE =
+        "FirestoreRecyclerAdapter"
+    private val FIRESTORE_REYCLER_OPTIONS_BUILDER_TYPE =
+        "FirestoreRecyclerOptions.Builder"
+
+    override fun visitClass(node: UClass) {
+      val adapterReferences = node
+          .fields
+          .filter { FIRESTORE_RECYCLER_ADAPTER_TYPE == it.type.canonicalText }
+          .map { AdapterReference(it) }
+
+      val recyclerOptionsReferences = node
+          .fields
+          .filter { FIRESTORE_REYCLER_OPTIONS_BUILDER_TYPE == it.type.canonicalText }
+          .toMutableSet()
+
+      node.accept(AdapterStartListeningMethodVisitor(adapterReferences))
+      node.accept(AdapterStopListeningMethodVisitor(adapterReferences))
+      node.accept(LifecycleOwnerMethodVisitor(recyclerOptionsReferences))
+
+      adapterReferences.forEach {
+        if (it.hasCalledStart && !it.hasCalledStop) {
+          context.report(
+              ISSUE_MISSING_LISTENING_STOP_METHOD,
+              it.uField,
+              context.getLocation(it.uField),
+              "Have called .startListening() without .stopListening()."
+          )
+        } else if (!it.hasCalledStart) {
```

Comment:
Does this trigger if `startListening` is not called but `setLifecycleOwner` is called?

Reply:
Nope have not taken that into account, but I was thinking it of it as well. I am not quite sure on the proper approach to solve this.

Currently `setLifecycleOwner` is applied on `FirestoreRecyclerOptions` which should be applied on `FirestoreRecyclerAdapter`, but there is nothing stopping the user of using both that method, as well as, the `start/stopListening` ones. So I am not quite sure how to prioritize those two together and chose to just add warnings based on the general usage and existence of lifecycle method invocations. That lets the user decide which warning to suppress, effectively acknowledging that it is the desired behavior and not just a method call that they forgot to invoke. 

@samtstern I am open to suggestions on the matter if you have anything to suggest :D

Reply:
Sure I can do that :)

Reply:
@samtstern I removed `setLifecyleOwner` related Lint checks and code :)